### PR TITLE
Blanczos randomized PCA

### PR DESCRIPTION
**Pull request by annamiraotoole**
- implements "Block Lanczos" / Blanczos algorithm for randomized SVD in Hail with HailTables
- adds method for Blanczos PCA following the exact interface of the current PCA method in Hail
- adds tests comparing Blanczos PCA with numpy and with Hail's current PCA

### CODE REVIEW

File: hail/python/hail/linalg/blanczos_data_script.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+def matmul_rowblocked_nonblocked(A, B):
+    temp = A.annotate_globals(mat = B)
+    temp = temp.annotate(ndarray = temp.ndarray @ temp.mat)
+    temp = temp.select(temp.ndarray)
+    temp = temp.drop(temp.mat)
+    return temp
+
+def matmul_colblocked_rowblocked(A, B):
+    temp = A.transmute(ndarray = block_product(A.ndarray.transpose(), B[A.row_group_number].ndarray))
+    result_arr_sum = temp.aggregate(block_aggregate(temp.ndarray))
```
Comment (johnc1231) [inline hail/python/hail/linalg/blanczos_data_script.py:123]: This code can now be rewritten as:

```
def matmul_colblocked_rowblocked(A, B):
    temp = A.transmute(ndarray = A.ndarray.transpose() @ B[A.row_group_number].ndarray)
    result_arr_sum = temp.aggregate(hl.agg.ndarray_sum(temp.ndarray)
    return result_arr_sum
```

which should allow you to get rid of `block_aggregate`, `block_product`, and `to_column_major` altogether.

File: hail/python/hail/methods/statgen.py [hunk 2]
```diff
@@ -1697,6 +1697,240 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+    def matmul_rowblocked_nonblocked(A, B):
+        temp = A.annotate_globals(mat = B)
+        temp = temp.annotate(ndarray = temp.ndarray @ temp.mat)
+        temp = temp.select(temp.ndarray)
+        temp = temp.drop(temp.mat)
+        return temp
+
+    def matmul_colblocked_rowblocked(A, B):
+        temp = A.transmute(ndarray = block_product(A.ndarray.transpose(), B[A.row_group_number].ndarray))
+        result_arr_sum = temp.aggregate(block_aggregate(temp.ndarray))
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1891]: Same comment as I made on the script, you no longer need the special aggregator.

File: hail/python/hail/linalg/blanczos_data_script.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    try:
+        table, mat = makeSharedData((3, n, m), block_size)
+    except Exception as e:
+        print(e)
+        print('failed to make data with ', (m, n))
+        return
+    
+    try:
+        G = hl.nd.array(np.random.normal(0, 1, (n,l)))
+        _, S, _, _, _, hail_time  = hailBlanczos(table, G, m, n, k, l, q, block_size)
```
Comment (johnc1231) [inline hail/python/hail/linalg/blanczos_data_script.py:224]: Is there a reason you don't just have this script call the new `hl._blanczos_pca`?

Comment (annamiraotoole) [reply hail/python/hail/linalg/blanczos_data_script.py:224]: I started writing it before I had the tests for the `_blanczos_pca` method working, I can change this now if I want. Though, I have timers and other useful data collection in this script, so I might want to keep it as well, if I don't want to change the return types of the hail method.

File: hail/python/hail/methods/statgen.py [hunk 4]
```diff
@@ -1697,6 +1697,244 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+    
+        # assert l > k
+        # assert n <= m
+        
+        Hi = matmul_rowblocked_nonblocked(A, G)
+        npH = concatToNumpy(Hi)
+        for j in range(0, q):
+            Hj = computeNextH(A, Hi)
+            npH = np.concatenate((npH, concatToNumpy(Hj)), axis=1)
+            Hi = Hj
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1915]: We can improve this loop. Right now, it does the following: 

Before the loop:
1. You multiply A by G once.
2. You collect that locally.

In the loop:
3. To compute Hj, you multiply AT by Hi, collect that down into a small matrix `nextG`, and then multiply A by `nextG`. 
4. Then you collect Hj locally.

We want something closer to Patrick's notes:

1. Start from `A`. It has one field, `ndarray`. 
2. Do:
```
temp = A.annotate(H_i = A.ndarray @ G)
temp = temp.annotate(G_i_intermediate = temp.ndarray.T @ temp.H_i)
result = temp.aggregate(hl.struct(Hi_chunks = hl.agg.collect(temp.H_i), G_i = hl.agg.ndarray_sum(temp.G_i))
localized_H_i = np.concat(result.Hi_chunks, axis = 1)
G_i = result.G_i
```
3. Append your `localized_H_i` to the `H` you're building up.
4. Repeat this process using the new `G_i`.

Comment (johnc1231) [reply hail/python/hail/methods/statgen.py:1915]: Note that A never changes, it's reused the same way every time.

File: hail/python/hail/methods/statgen.py [hunk 5]
```diff
@@ -1697,6 +1697,244 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+        block = prod.block
+        return hl.nd.from_column_major(
+            hl.agg.array_sum(block),
+            hl.agg.take(shape, 1)[0])
+
+    # returns flat array
+    def to_column_major(ndarray):
+        n_rows, n_cols = ndarray.shape
+        return hl.range(hl.int(n_rows * n_cols)).map(
+            lambda absolute: ndarray[absolute % n_rows, absolute // n_rows])
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1880]: Delete these, no longer used.

File: hail/python/hail/methods/statgen.py [hunk 6]
```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+        blocks = A.ndarray.collect()
+        big_mat = np.concatenate(blocks, axis=0)
+        return big_mat
+
+    def ndarray_to_table(chunked_arr):
+        structs = [hl.struct(row_group_number = hl.int64(idx), ndarray = block)
+                   for idx, block in enumerate(chunked_arr)]
+        ht = hl.Table.parallelize(structs)
+        ht = ht.key_by('row_group_number')
+        return ht
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1866]: All no longer used, can delete.

File: hail/python/hail/linalg/blanczos_data_script_new.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import numpy as np
```
Comment (johnc1231) [inline hail/python/hail/linalg/blanczos_data_script_new.py:1]: At some point, remove this whole script, we aren't going to check in test scripts.

File: hail/python/hail/linalg/blanczos_data_script_old.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
+import numpy as np
```
Comment (johnc1231) [inline hail/python/hail/linalg/blanczos_data_script_old.py:1]: At some point, remove this whole script, we aren't going to check in test scripts.

Comment (johnc1231) [reply hail/python/hail/linalg/blanczos_data_script_old.py:1]: Also remove the notebooks.

File: hail/python/hail/methods/statgen.py [hunk 9]
```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+    info("hwe_normalized_pca: running PCA using {} variants.".format(n_variants))
+
+    mt = mt.annotate_rows(__mean_gt=mt.__AC / mt.__n_called)
+    mt = mt.annotate_rows(
+        __hwe_scaled_std_dev=hl.sqrt(mt.__mean_gt * (2 - mt.__mean_gt) * n_variants / 2))
+    mt = mt.unfilter_entries()
+
+    normalized_gt = hl.or_else((mt.__gt - mt.__mean_gt) / mt.__hwe_scaled_std_dev, 0.0)
+
+    return pca(normalized_gt,
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:2014]: Needs to call `_blanczos_pca`.

File: hail/python/hail/methods/statgen.py [hunk 10]
```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+    lt = ht.select()
+    lt = lt.annotate_globals(US = us)
+    lt = lt.add_index()
+    idx = lt.key
+    lt = lt.annotate(loadings = lt.US[lt.idx,:]._data_array())
+
+    if compute_loadings:
+        return eigens, st, lt
+    else:
+        return eigens, st
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1953]: This is how numpy would handle a situation like this, but Hail PCA currently always returns a 3-tuple, with `None` as the third element if `compute_loadings` is false.

File: hail/python/hail/methods/statgen.py [hunk 11]
```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+    hail_scores = hl.nd.array(scores)
+    hail_array_scores = hail_scores._data_array()
+    cols_and_scores = hl.zip(ht.index_globals().cols, hail_array_scores).map(lambda tup: tup[0].annotate(scores = tup[1]))
+    st = hl.Table.parallelize(cols_and_scores, key=list(mt.col_key))
+
+    us = hl.nd.array(U @ matrix_S)
+    lt = ht.select()
+    lt = lt.annotate_globals(US = us)
+    lt = lt.add_index()
+    idx = lt.key
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1947]: Delete this line, doesn't do anything.

File: hail/python/hail/methods/statgen.py [hunk 12]
```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+
+    scores = V.transpose() @ matrix_S
+    eigens = S * S
+
+    hail_scores = hl.nd.array(scores)
+    hail_array_scores = hail_scores._data_array()
+    cols_and_scores = hl.zip(ht.index_globals().cols, hail_array_scores).map(lambda tup: tup[0].annotate(scores = tup[1]))
+    st = hl.Table.parallelize(cols_and_scores, key=list(mt.col_key))
+
+    us = hl.nd.array(U @ matrix_S)
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1943]: Don't scale U.

File: hail/python/hail/methods/statgen.py [hunk 13]
```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
...
+            
+        if times:
+            return truncV, truncS, truncW, q_loop_time, process_Q_time, svd_time, get_V_time
+        else:
+            return truncV, truncS, truncW
+
+    U, S, V = hailBlanczos(A, G, k, l, q, block_size, report_times)
+    matrix_S = np.diag(S)
+
+    scores = V.transpose() @ matrix_S
```
Comment (johnc1231) [inline hail/python/hail/methods/statgen.py:1935]: Get rid of `matrix_S`. No reason to manifest the matrix. Faster to do `scores = V.transpose() * S`, gives same result.

**Comment by johnc1231** [comment]
Ok, merged to resolve conflicts. Going to now add a few commits to fix errors.

**Pull request merged**